### PR TITLE
Revert irept change that stopped sharing and add unit test

### PR DESCRIPTION
**Pull request by tautschnig**
This is the test from #1855 provided by @NathanJPhillips, with the modification to only make it run when `SHARING` is enabled.

### CODE REVIEW

**Review by peterschrammel — APPROVED** [review]

**Comment by kroening** [comment]
This seems to do more than the title suggests?

**Comment by tautschnig** [comment]
The title might just have been very long and probably had priorities wrong - it's changed now.

**Comment by tautschnig** [comment]
@NathanJPhillips any objections against this one? @smowton is out for the week; anyone else who wants to jump on this one?

**Review by kroening — APPROVED** [review]

**Comment by kroening** [comment]
I should stress that I am happy to re-contemplate using C++ standard library idioms for irept, with a sufficiently strong case that it'll be robust.

**Comment by NathanJPhillips** [comment]
Adding a test and then immediately commenting it out because you know that it would fail seems like a red flag to me. Reuben's change may have been more conservative that it needed to be, but that is better than allowing incorrect behaviour. If you find a bug then it is fine to fix it straight away and then apply optimisations later: incremental is good. The performance impact of Reuben's fix was already analysed and accepted at the time it was applied. Reverting the fix would not be a good solution even if we didn't have the optimisation already in review.

Saying that `irept` was good for 15 years ignores the fact that we are constantly finding and fixing bugs in CBMC. As we exercise the product on more challenging tasks it is not good enough just to have it work as well as it used to.

**Comment by tautschnig** [comment]
@NathanJPhillips  Sentence-by-sentence response:
> Adding a test and then immediately commenting it out because you know that it would fail seems like a red flag to me.

Ok, but it seems better than not having a test?

> Reuben's change may have been more conservative that it needed to be, but that is better than allowing incorrect behaviour.

No, performance and memory consumption matter. "Allowing" incorrect behaviour is different from exhibiting it.

> If you find a bug then it is fine to fix it straight away and then apply optimisations later: incremental is good.

Incrementally following the wrong route permits backtracking, which is what this PR does.

> The performance impact of Reuben's fix was already analysed and accepted at the time it was applied. 

Which simply shows that my performance analysis was not good enough.

> Reverting the fix would not be a good solution even if we didn't have the optimisation already in review.

Why? On the contrary, allowing people to use a broken solution is not a good solution.

> Saying that `irept` was good for 15 years ignores the fact that we are constantly finding and fixing bugs in CBMC.

The way you are conflating one data structure with the entire code base severely undermines the credibility of what you are saying.

> As we exercise the product on more challenging tasks it is not good enough just to have it work as well as it used to.

I don't know what the size of your experimental corpus is. Mine is the >400 million lines of C code in Debian.

**Comment by NathanJPhillips** [comment]
I have no disagreements with @tautschnig response. If people aren't convinced by my earlier comment then go ahead with this revert.

File: src/util/irep.h [hunk 1]
```diff
@@ -122,6 +122,7 @@ class irept
   {
     if(data!=&empty_d)
     {
+      // NOLINTNEXTLINE(build/deprecated)
       assert(data->ref_count!=0);
```
Comment (martin-cs) [inline src/util/irep.h]: I'd be kinda happier if this was an invariant for the sake of uniformity.  I was going to add a comment along the lines of "but everywhere it is used also uses std_expr.h or std_types.h so it won't make much difference" but that's not true (shows how often I wander through the front-end!).  Not exactly sure of the impact on build speed but ... I guess we can put off converting this until we've done things like minimising the use of headers to improve build speed.

**Review by martin-cs — APPROVED** [review]
Approve because I believe this to be a correct implementation of : https://github.com/diffblue/cbmc/pull/1855#issuecomment-366431321

FWIW I have noticed a small but significant slow-down in a key use-case today.  It was something that was added in the past ... 6-9 months and must be pretty core.  I'll try with this patch and see if it changes anything.

File: src/util/irep.h (hunk 1 above)
Comment (tautschnig) [reply src/util/irep.h]: I am very open to resolving this in a different way, i.e., by actually using `INVARIANT`. I just feared some undesirable impact.

**Comment by tautschnig** [comment]
>  I'll try with this patch and see if it changes anything.

@martin-cs, if you have any results on this then please let me know.

**Comment by tautschnig** [comment]
This is currently failing tests merged in #1943 and I am thus debugging this.

**Comment by tautschnig** [comment]
Looking at #1964 (documentation changes only), the problem isn't actually caused by the changes on this PR.

**Comment by tautschnig** [comment]
Marking this as do-not-merge as #1970 ought to be merged first and experiments to come up with memory regression tests are in progress.

**Comment by kroening** [comment]
This one seems to be essential before we do the release.

**Comment by tautschnig** [comment]
On top of 63acc5bedd16481368f4760b414bf831afbbd4aa on a customer-provided
benchmark:
```
size of program expression: 127981 steps
slicing removed 16144 assignments
Generated 1 VCC(s), 1 remaining after simplification
Passing problem to refinement loop with MiniSAT 2.2.1 with simplifier
converting SSA
Running refinement loop with MiniSAT 2.2.1 with simplifier
BV-Refinement: post-processing
BV-Refinement: iteration 1
1014336 variables, 4131455 clauses
SAT checker inconsistent: instance is UNSATISFIABLE
BV-Refinement: got UNSAT, and the proof passes => UNSAT
Total iterations: 1
Runtime decision procedure: 1.84756s
```

User time (seconds): 16.42
User time (seconds): 16.43
User time (seconds): 16.41

#1855 + #2316
User time (seconds): 14.14
User time (seconds): 13.97
User time (seconds): 14.03
86633882 calls to small_shared_ptrt<irept::dt>::~small_shared_ptrt()

#1860 + #2316
User time (seconds): 12.39
User time (seconds): 12.57
User time (seconds): 12.64

#2316
User time (seconds): 14.66
User time (seconds): 14.52
User time (seconds): 14.38
105950686 calls to small_shared_ptrt<irept::dt>::~small_shared_ptrt()

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: cbce4bc11741b4310a16b2ef3547ea6885da591a).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — CHANGES_REQUESTED** [review]
This needs an update PR to test-gen.

**Comment by tautschnig** [comment]
What is it that this PR breaks in test-gen? That's a bit unexpected, except if it is about missing include files (which should likely be fixed either way).

**Comment by allredj** [comment]
Seems to be a missing include, indeed.

**Comment by allredj** [comment]
Test-gen update is ready for review: https://github.com/diffblue/test-gen/pull/1941

**Comment by allredj** [comment]
Sorry the test-gen update exhibited a another bug. I'll try to fix all that as quickly as possible, but that might not be done today...

**Comment by kroening** [comment]
This is getting urgent, as blocking release!

**Comment by allredj** [comment]
The submodule update https://github.com/diffblue/test-gen/pull/1941 has now been rebase on the fix. Waiting for CI.

**Comment by allredj** [comment]
@tautschnig could you please rebase on cbmc/develop?

**Comment by allredj** [comment]
@kroening this needs a rebase on cbmc/develop.

**Pull request merged**